runtimevar/filevar/filevar.go

```diff
@@ -111,68 +111,58 @@ func (w *watcher) WatchVariable(ctx context.Context) (driver.Variable, error) {
 		// Continue.
 	}
 
-	// Detect if there has been a changed first. If there are no changes, then proceed to add a
-	// fsnotify watcher.
-	wait := true
-	for wait {
-		sd := w.processFile()
-		switch sd.state {
-		case errorState:
-			return zeroVar, sd.err
+	// Start watching over the file and wait for events/errors.
+	// We'll get a notifierErr if the file doesn't currently exist.
+	// Also, note that we may never use the notifier if there's already
+	// a change to the file. We must create it now before checking to
+	// avoid race conditions.
+	notifierErr := w.notifier.Add(w.file)
+	if notifierErr == nil {
+		defer w.notifier.Remove(w.file)
+	}
 
-		case hasChangedState:
-			return *sd.variable, nil
+	for {
+		if v, err := w.processFile(); err != nil {
+			return zeroVar, err
+		} else if v != nil {
+			return *v, nil
+		}
 
-		case stillDeletedState:
-			// Last known state is deleted, need to wait for file to show up before adding a watch.
+		if w.isDeleted {
+			// Last known state is deleted, need to wait for file to show up.
 			t := time.NewTimer(w.waitTime)
 			select {
 			case <-t.C:
 			case <-ctx.Done():
 				t.Stop()
 				return zeroVar, ctx.Err()
 			}
-
-		case noChangeState:
-			wait = false
+			continue
 		}
-	}
-
```

Comment:
This is where the race was before; if you insert a Sleep(1 * time.Second) here some of the tests hang forever. It happens if the change occurs after the processFile above, but before the notifier.Add happens, so the notifier never gets an event.